encryption_jwt/codec_server.py

```diff
@@ -0,0 +1,119 @@
+import os
+import ssl
+import logging
+import jwt
+import grpc
+from aiohttp import hdrs, web
+
+from temporalio.api.common.v1 import Payload, Payloads
+from temporalio.api.cloud.cloudservice.v1 import request_response_pb2, service_pb2_grpc
+from google.protobuf import json_format
+from encryption_jwt.codec import EncryptionCodec
+
+AUTHORIZED_ACCOUNT_ACCESS_ROLES = ["admin"]
+AUTHORIZED_NAMESPACE_ACCESS_ROLES = ["read", "write", "admin"]
+
+temporal_ops_address = "saas-api.tmprl.cloud:443"
+if os.environ.get("TEMPORAL_OPS_ADDRESS"):
+    temporal_ops_address = os.environ.get("TEMPORAL_OPS_ADDRESS")
+
+
+def build_codec_server() -> web.Application:
+    # Cors handler
+    async def cors_options(req: web.Request) -> web.Response:
+        resp = web.Response()
+
+        if req.headers.get(hdrs.ORIGIN) == "http://localhost:8080":
+            logger.info("Setting CORS headers for localhost")
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_ORIGIN] = "http://localhost:8080"
+
+        elif req.headers.get(hdrs.ORIGIN) == "https://cloud.temporal.io":
+            logger.info("Setting CORS headers for cloud.temporal.io")
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_ORIGIN] = "https://cloud.temporal.io"
+
+        allow_headers = "content-type,x-namespace"
+        if req.scheme.lower() == "https":
+            allow_headers += ",authorization"
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_CREDENTIALS] = "true"
+
+        # common
+        resp.headers[hdrs.ACCESS_CONTROL_ALLOW_METHODS] = "POST"
+        resp.headers[hdrs.ACCESS_CONTROL_ALLOW_HEADERS] = allow_headers
+
+        return resp
+
+    def decryption_authorized(email: str, namespace: str) -> bool:
+        credentials = grpc.composite_channel_credentials(grpc.ssl_channel_credentials(
+        ), grpc.access_token_call_credentials(os.environ.get("TEMPORAL_API_KEY")))
+
+        with grpc.secure_channel(temporal_ops_address, credentials) as channel:
+            client = service_pb2_grpc.CloudServiceStub(channel)
```

Comment:
@cretz is there an example of using this? I was trying trying to get it to work but running into errors and some of the types aren't documented yet, so it was tough to figure out what I'm doing wrong.

Reply:
No, it is still experimental and this may not be its final form (and even the gRPC service could change, it is not GA yet). It's basically direct gRPC similar to what you're doing. Here is a test case where we call it: https://github.com/temporalio/sdk-python/blob/2e6d742ce1531741c0f953a311689bd72781b720/tests/test_client.py#L1315-L1325.

Reply:
Thanks! I got this working 👍